src/hooks/unstable_useSlots.ts

```diff
@@ -0,0 +1,65 @@
+import React from 'react'
+
+type SlotConfig = Record<string, React.ComponentType>
+
+type SlotElements<Type extends SlotConfig> = {
+  [Property in keyof Type]: React.ReactElement
+}
+
+/**
+ * Extract components from `children` so we can render them in different places,
+ * allowing us to implement components with SSR-compatible slot APIs.
+ * Note: We can only extract direct children, not nested ones.
+ */
+// eslint-disable-next-line camelcase
+export function unstable_useSlots<T extends SlotConfig>(
+  children: React.ReactNode,
+  config: T,
+): [Partial<SlotElements<T>>, React.ReactNode[]] {
+  // Object mapping slot names to their elements
+  const slots: Partial<SlotElements<T>> = mapValues(config, () => undefined)
```

Comment:
Since this is a hot path in render, do you think it would help to reduce the object transforms in `mapValues` and `getKeyByValue`? Like for keys, reduce, find, etc. Or would the different not really matter, ultimately?

One quick idea was to generate arrays for keys and value of `config` and use them to index each other:

```tsx
const slots = {};
const keys = Object.keys(config);
const values = Object.values(config);
const rest = [];

React.Children.forEach(children, child => {
  if (!React.isValidElement(child)) {
    rest.push(child);
    return;
  }

  const index = values.findIndex(value => {
    return child.type === value;
  });

  if (index === -1) {
    rest.push(child);
    return;
  }

  const slotKey = keys[index];
  if (slots[slotKey]) {
    return;
  }

  slots[slotKey] = child;
});
```

Another idea could be to use `Object.entries()` along with `find()` if that would read better.

Reply:
Love it! This is wayy faster:

<img width="861" alt="CleanShot 2023-03-15 at 18 24 57@2x" src="https://user-images.githubusercontent.com/4608155/225477857-2c596e2f-4494-4872-8619-8c3f442f8c9d.png">